src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+{
+    return QModelIndex();  // Tags are single-level.
+}
+
+int TagFilterModel::rowCount(const QModelIndex &parent) const
+{
+    if (!parent.isValid())
+        return m_tagRows.count();
+    return 0;
+}
+
+QModelIndex TagFilterModel::index(const QString &tag) const
+{
+    return index(findItem(tag));
+}
+
+QString TagFilterModel::tag(const QModelIndex &index) const
+{
+    if (!index.isValid()) return QString();
+    return static_cast<TagModelItem *>(index.internalPointer())->tag();
+}
+
+QModelIndex TagFilterModel::index(const TagModelItem *item) const
+{
+    if (!item)
+        return QModelIndex();
+    const int row = m_tagRows.indexOf(item->tag());
+    if (row < 0)
+        return QModelIndex();
+    return index(row, 0, QModelIndex());
+}
+
+void TagFilterModel::tagAdded(const QString &tag)
+{
+    const int row = m_tagRows.count();
+    beginInsertRows(QModelIndex(), row, row);
+    addItemToModel(new TagModelItem(tag, tag));
+    endInsertRows();
+}
+
+void TagFilterModel::tagRemoved(const QString &tag)
+{
+    TagModelItem *item = findItem(tag);
+    if (item) {
+        QModelIndex i = index(item);
+        beginRemoveRows(i.parent(), i.row(), i.row());
+        removeItemFromModel(item);
+        endRemoveRows();
+    }
+}
+
+void TagFilterModel::torrentTagAdded(BitTorrent::TorrentHandle *const torrent, const QString &tag)
+{
+    if (torrent->tags().count() == 1)
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->increaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentTagRemoved(BitTorrent::TorrentHandle* const torrent, const QString &tag)
+{
+    Q_ASSERT(torrent->tags().count() >= 0);
+    if (torrent->tags().count() == 0)
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->decreaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentAdded(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->increaseTorrentsCount();
+
+    const QVector<TagModelItem *> items = findItems(torrent->tags());
+    if (items.isEmpty())
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    foreach (TagModelItem *item, items)
+        item->increaseTorrentsCount();
+}
+
+void TagFilterModel::torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->decreaseTorrentsCount();
+
+    if (torrent->tags().isEmpty())
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    foreach (TagModelItem *item, findItems(torrent->tags()))
+        item->decreaseTorrentsCount();
+}
+
+QString TagFilterModel::getSpecialAllTag()
+{
+    static const QString *const ALL_TAG = new QString(" ");
```

Comment:
https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2

This is the second question I've gotten on this, so if people feel this idiom is too obscure and hurts readability, we can get rid of the static and just create a new string object on every call.